sched/semaphore/sem_wait.c

```diff
@@ -63,8 +63,9 @@
  *   returned on success.  A negated errno value is returned on failure.
  *   Possible returned errors:
  *
- *   - EINVAL:  Invalid attempt to get the semaphore
- *   - EINTR:   The wait was interrupted by the receipt of a signal.
+ *   - EINVAL:   Invalid attempt to get the semaphore
+ *   - EINTR:    The wait was interrupted by the receipt of a signal.
+ *   - ETIMEOUT: The wait was interrupted by the receipt of a timeout signal.
```

Comment:
I do not support this change. It changes the behavior of a core semaphore object in favor of a partial implementation. In this case the `uninterruptible` waiting for semaphore does not make any sense. Despite there is an attempt to cover `pthread_rwlock` implementation only by checking `if (info->si_signo == SIGCONDTIMEDOUT)`, but this code does not distinguish between `pthread_cond_timedwait` and `pthread_rwlock_clockrdlock`/`pthread_rwlock_clockwrlock`. Also the user may simply choose the same signo value for other part of an application that does not use pthreads and this code will introduce unexpected side effects.

PR description states: "Upon return from the signal handler the thread shall resume waiting as if it is not interrupted.", but that goal is not achieved in this PR because the same `ts` value is used if `EINTR` condition is detected. I recommend to solve it in another way. Calculate the number of ticks elapsed from the start of `pthread_cond_clockwait()` till the point when `EINTR` is detected and resume waiting only for the remaining time if that is a positive value. In such way changes can be localized at `pthread_rwlock` level instead of changing the behavior of the core synchronization primitive.

Reply:
Upd: I missed that `ts` is an absolute timeout, so using `continue` is acceptable. So most probably we can just change
```
                  /* Start the watchdog */

                  wd_start(&rtcb->waitdog, ticks,
                           pthread_condtimedout, mypid);

                  /* Take the condition semaphore.  Do not restore
                   * interrupts until we return from the wait.  This is
                   * necessary to make sure that the watchdog timer and
                   * the condition wait are started atomically.
                   */

                  status = nxsem_wait((FAR sem_t *)&cond->sem);

                  /* We no longer need the watchdog */

                  wd_cancel(&rtcb->waitdog);

                  /* Did we get the condition semaphore. */

                  if (status < 0)
                    {
                      /* NO.. Handle the special case where the semaphore
                       * wait was awakened by the receipt of a signal --
                       * presumably the signal posted by
                       * pthread_condtimedout().
                       */

                      if (status == -EINTR)
                        {
                          ret = clock_abstime2ticks(clockid, abstime, &ticks);
                          
                          swarn("WARNING: Timedout!\n");
                          ret = ETIMEDOUT;
                        }
                      else
                        {
                          ret = status;
                        }
                    }
```
to
```
                  /* Start the watchdog */

                  wd_start(&rtcb->waitdog, ticks,
                           pthread_condtimedout, mypid);

                  /* Take the condition semaphore.  Do not restore
                   * interrupts until we return from the wait.  This is
                   * necessary to make sure that the watchdog timer and
                   * the condition wait are started atomically.
                   */

                  while ((status = nxsem_wait((FAR sem_t *)&cond->sem)) < 0)
                    {
                      /* Handle the special case where the semaphore
                       * wait was awakened by the receipt of a signal --
                       * presumably the signal posted by
                       * pthread_condtimedout().
                       */

                      if (status == -EINTR)
                        {
                          if (wd_gettime(&rtcb->waitdog) <= 0)
                            {
                              swarn("WARNING: Timedout!\n");
                              ret = ETIMEDOUT;
                              break;
                            }
                        }
                      else
                        {
                          ret = -status;
                          break;
                        }
                    }

                  /* We no longer need the watchdog */

                  wd_cancel(&rtcb->waitdog);
```
should be sufficient to fix the issue

Reply:
But why not issue set the error code correctly in the first place?

Reply:
What do you mean?

Reply:
As I read the problem description: we need to distinguish if lock wait was interrupted by a timeout or by another signal. If it is not a timeout signal, then we need to continue waiting, isn't it?

Reply:
> As I read the problem description: we need to distinguish if lock wait was interrupted by a timeout or by another signal. If it is not a timeout signal, then we need to continue waiting, isn't it?

,Yes. I mean: since nxsig_tcbdispatch can distinguish the wakeup reason: timeout or interrupt, nxsig_tcbdispatch should set the correct error code before wake up the task, instead the waken task check the timeout.

Reply:
As `SIGCONDTIMEDOUT ` is published from `pthread_condtimedout` as I proposed above we can just `if (wd_gettime(&rtcb->waitdog) <= 0)` to check for timeout event.

Reply:
The original purpose of this patch was to solve inconsistency between nuttx pthread_rwlock_timedwrlock() and posix standard.
In the progress I found that the reason was that the original code expected nxsem_wait to return ETIMEOUT.
So Regardless of which implemention I think we should set nxsem_wait() error_return correctly.

Reply:
@pkarashchenko please check:
https://github.com/apache/incubator-nuttx/blob/master/sched/semaphore/sem_tickwait.c#L67
https://github.com/apache/incubator-nuttx/blob/master/sched/semaphore/sem_tickwait.c#L128
You can see not only pthread_rwlock_timedwrlock, but also sem_timedwait has the similar issue.

Reply:
What I'm trying to point out that if we rely on `if (info->si_signo == SIGCONDTIMEDOUT)` we can't limit only `pthread_condtimedout` to publish that signo. We can't restrict user not to `kill(pid, 16)` (16 is default value for `SIGCONDTIMEDOUT`). And with this change we will consider `kill(pid, 16)` will be considered as a timeout event for rwlock

Reply:
Ok, please wait @jihandong recheck internally.

Reply:
Maybe we can even eliminate `pthread_condtimedout` and use semaphore non-interruptable timed wait here? Please analyze if that can be applied. If yes, then it will solve all the proper error code handling.

Reply:
We also have `pthread_sem_take` that is used in `pthread_cond_wait` and most probably is the best candidate to be used here.

Reply:
> Maybe we can even eliminate `pthread_condtimedout` and use semaphore non-interruptable timed wait here? Please analyze if that can be applied. If yes, then it will solve all the proper error code handling.

This way can solve the problem well.